### PR TITLE
[Try] Ensure that all image tags run through `wp_filter_content_tags`

### LINKED ISSUE #7981: Blocks/Renderer: Consider lazy-loading images
## Feature Description

[As of](https://make.wordpress.org/core/2020/07/14/lazy-loading-images-in-5-5/) WordPress 5.5, all images have the loading="lazy" attribute added. This has a number of preference benefits not to mention resource saving. 

There are a number of places where raw image tags at output. See 

 https://github.com/google/web-stories-wp/blob/a31592468d3145c266a2577692f53a0aeb028a10/includes/Renderer/Story/Image.php#L83-L89

https://github.com/google/web-stories-wp/blob/823cd1af97c5f124948758d429e63d32ccb90667/includes/Renderer/Stories/Renderer.php#L571-L576

We should ensure that these image tags are out with the lazying loading attribute, while making sure that compile with the `wp_img_tag_add_loading_attr` filter. This could be done by sampling passing the generated image tag through the [wp_img_tag_add_loading_attr](https://developer.wordpress.org/reference/functions/wp_img_tag_add_loading_attr/) function. 

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Added in WordPress 5.5, the `wp_filter_content_tags` function, adds the source set and lazy attributes to images. If those attributes are controlled by filters and developer out of our there use. This PR tries to ensure that all image tags run through this function and see if images get the correct attributes. 

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7981

### CODE REVIEW

File: includes/Renderer/Story/Image.php [hunk 1]
```diff
@@ -81,11 +82,12 @@ public function render( array $args = [] ) {
...
 				if ( ! empty( $this->story->get_poster_portrait() ) ) {
 					printf(
-						'<img src="%1$s" width="%2$d" height="%3$d" alt="%4$s" />',
+						'<img src="%1$s" width="%2$d" height="%3$d" alt="%4$s" class="%5$s" />',
 						esc_url( $this->story->get_poster_portrait() ),
 						absint( $args['width'] ),
 						absint( $args['height'] ),
-						esc_attr( $this->story->get_title() )
+						esc_attr( $this->story->get_title() ),
+						esc_attr( 'wp-image-' . $this->story->get_thumbnail_id() )
```
Comment (spacedmonkey) [inline includes/Renderer/Story/Image.php:89]: This attribute is required. See this block. 

https://github.com/WordPress/wordpress-develop/blob/e2feb7cf77603dcb0622b105377be7a37ffabf99/src/wp-includes/media.php#L1789-L1821

File: tests/phpunit/tests/Renderer/Stories/Renderer.php [hunk 2]
```diff
@@ -158,15 +169,28 @@ public function test_render_story_with_poster() {
...
+		$this->story_query->method( 'get_stories' )->willReturn(
+			[ $this->story_model ]
+		);
+
 		$renderer = $this->getMockForAbstractClass( AbstractRenderer::class, [ $this->story_query, $this->assets, $this->register_global_assets ], '', true, true, true, [ 'is_amp_request' ] );
 		$renderer->expects( $this->any() )->method( 'is_amp_request' )->willReturn( false );
+		$this->call_private_method( $renderer, 'init' );
 		$this->set_private_property( $renderer, 'stories', [ $this->story_model ] );
+		$this->set_private_property( $renderer, 'height', 300 );
+		$this->set_private_property( $renderer, 'width', 300 );
```
Comment (spacedmonkey) [inline tests/phpunit/tests/Renderer/Stories/Renderer.php:181]: Force the image size to a square, so we can load the srcset attribute.

File: tests/phpunit/tests/Renderer/Stories/Renderer.php [hunk 3]
```diff
@@ -84,7 +95,7 @@ public static function wpSetUpBeforeClass( $factory ) {
 	public function setUp() {
 		parent::setUp();
 
-		$this->story_model = $this->createMock( Story::class );
+		$this->story_model = new \Google\Web_Stories\Model\Story();
```
Comment (spacedmonkey) [inline tests/phpunit/tests/Renderer/Stories/Renderer.php:98]: There is no need to mock this here.

File: tests/phpunit/tests/Renderer/Stories/Renderer.php [hunk 4]
```diff
@@ -76,6 +76,17 @@ public static function wpSetUpBeforeClass( $factory ) {
...
+		$poster_attachment_id = $factory->attachment->create_object(
+			[
+				'file'           => DIR_TESTDATA . '/images/canola.jpg',
+				'post_parent'    => 0,
+				'post_mime_type' => 'image/jpeg',
+				'post_title'     => 'Test Image',
+			]
+		);
+		wp_maybe_generate_attachment_metadata( get_post( $poster_attachment_id ) );
+		set_post_thumbnail( self::$story_id, $poster_attachment_id );
```
Comment (spacedmonkey) [inline tests/phpunit/tests/Renderer/Stories/Renderer.php:88]: Improve tests.

File: includes/Renderer/Story/Image.php [hunk 5]
```diff
@@ -52,6 +52,7 @@ class Image {
 	 */
 	public function __construct( Story $story ) {
 		$this->story = $story;
+		add_filter( 'web_stories_render_image', 'wp_filter_content_tags' );
```
Comment (spacedmonkey) [inline includes/Renderer/Story/Image.php]: There maybe a better place for this....

Comment (swissspidy) [reply includes/Renderer/Story/Image.php]: Yeah we should not be adding filters in constructors.

**Comment by swissspidy** [comment]
So this attempts to fix #7981 **_and_** #7983? If so, can you update the PR description in that case (One `Fixes #...` per line) & connect them in ZenHub?

As mentioned in #7981, not all images should have `loading=lazy`, so we'd need a way to opt out of that in some cases. How can we do that?

Also, can't we just use `wp_get_attachment_image()` in these places instead of doing output buffering?

**Comment by spacedmonkey** [comment]
Sorry this is me trying different things to try and fix the problem. I think is is a problem that our image tags do not go through all the filter and action that `wp_filter_content_tags` function adds. 

As for wp_get_attachment_image, that would mean storing the output of this function at the story model level. We can do this, but I am not sure it is the best idea.

**Comment by swissspidy** [comment]
> As for wp_get_attachment_image, that would mean storing the output of this function at the story model level. We can do this, but I am not sure it is the best idea.

Or just storing the poster attachment ID on the model? Wouldn't that work?

File: includes/Renderer/Stories/Renderer.php [hunk 6]
```diff
@@ -567,17 +568,31 @@ protected function render_story_with_poster() {
...
+				} else {
+
+					printf(
+						'<img src="%1$s" width="%2$d" height="%3$d" alt="%4$s" class="%5$s" />',
+						esc_url( $story->get_poster_portrait() ),
+						absint( $this->width ),
+						absint( $this->height ),
+						esc_attr( $story->get_title() ),
+						esc_attr( 'wp-image-' . $story->get_thumbnail_id() )
+					);
```
Comment (swissspidy) [inline includes/Renderer/Stories/Renderer.php:579]: IMO it was more readable before. Plus, we don't use `printf` for `amp-img` above.

Can we use the  previous way of writing `` again for consistency?

File: includes/Renderer/Stories/Renderer.php [hunk 7]
```diff
@@ -567,17 +568,31 @@ protected function render_story_with_poster() {
...
-						width="<?php echo absint( $this->width ); ?>"
-						height="<?php echo absint( $this->height ); ?>"
-					>
-				<?php } ?>
+					<?php
+				} else {
+
+					printf(
+						'<img src="%1$s" width="%2$d" height="%3$d" alt="%4$s" class="%5$s" />',
+						esc_url( $story->get_poster_portrait() ),
```
Comment (swissspidy) [inline includes/Renderer/Stories/Renderer.php:574]: Could just use `$poster_url` as before

**Comment by swissspidy** [comment]
Overall looks good 👍

**Comment by swissspidy** [comment]
Closing since this is stale.

**Pull request closed without merging**